WordPress/gutenberg
# CustomGradientBar: Fix insertion and control point positioning to more closely follow cursor

## What?


Fixes part of https://github.com/WordPress/gutenberg/issues/41390

This is a pretty subtle issue to look at visually, but it's much more noticeable when interacting with the component.

Update the `CustomGradientBar` component to fix the positioning of the control points and inserter so that it follows the cursor position more accurately.

## Why?


As raised in #41390, the existing logic for where to position the control points and inserter in the custom gradient bar is slightly off — it's more noticeable at larger viewport sizes when the component is rendered in Storybook, however it's still a bit fiddly in real world usage in the gradient color picker in the editor, so this PR should hopefully help improve the UX of interacting with the controls.

## How?


* Simplify the logic of how we calculate the control point and insertion points so that their position is always at the center of the rendered icon.
* Then, use a `calc()` rule when rendering the position to offset the control point so that it's centered in its position.
* Update the CSS so that instead of relying on a `padding-right` setting, we use a container with slightly less width.
* Move the `ref` for tracking the positions of control points to the container of the markers, to improve accuracy (this is the component that the markers are positioned in relation to, so I _think_ this should help with the accuracy, too)
* Add tests for the calculation.

## Testing Instructions





1. Test using the component in Storybook:
  - Run `npm run storybook:dev` 
  - Then, visit: http://localhost:50240/?path=/story/components-customgradientpicker--default
2. Use the CustomGradientPicker component in the editor, and check that interacting with the control and insertion points feels good

Check that tests pass (they can be run locally via: `npm run test-unit -- --testPathPattern packages/components/src/custom-gradient-bar/test/utils.js`)

## Screenshots or screencast 

| Before | After |
| --- | --- |
| ![2022-06-02 16 14 45](https://user-images.githubusercontent.com/14988353/171565197-c3c5b2bd-49cf-415c-836a-ddf2b08fb18d.gif) | ![2022-06-02 16 13 20](https://user-images.githubusercontent.com/14988353/171565221-1dc384b1-1cd0-4bd0-acbb-7a0e63b52b45.gif) |
| ![2022-06-02 16 02 46](https://user-images.githubusercontent.com/14988353/171563699-f573671c-4cd7-4ed9-a5dc-b3b90916fbff.gif) | ![2022-06-02 16 06 09](https://user-images.githubusercontent.com/14988353/171564001-7539efeb-76e5-4115-928c-7f05cc2d48bf.gif) |
